### PR TITLE
Deinterlacing improvements

### PR DESCRIPTION
**Pull request by orryverducci**
**Part 1:** This
**Part 2:** https://github.com/jellyfin/jellyfin-web/pull/1798

This is my first contribution to Jellyfin so all feedback is welcome!

This makes a few small improvements to Jellyfin's deinterlacing functionality while transcoding media:

- Adds the [bwdif](https://ffmpeg.org/ffmpeg-filters.html#bwdif) deinterlacer from FFmpeg as a deinterlace filter option
  - I've left yadif as the default, but bwdif is higher quality than yadif (less artifacts) and similar in speed, so it's worth considering for this to be made the default
- Adds an option to enable double frame rate (bob) deinterlacing supported by both the software and hardware deinterlacers (excluding Quick Sync due to a FFmpeg limitation)
- Slightly simplifies the software deinterlacing filter building code

### CODE REVIEW

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 1]
```diff
@@ -2088,6 +2088,9 @@ private string GetFixedSizeScalingFilter(Video3DFormat? threedFormat, int reques
             var hasTextSubs = state.SubtitleStream != null && state.SubtitleStream.IsTextSubtitleStream && state.SubtitleDeliveryMethod == SubtitleDeliveryMethod.Encode;
             var hasGraphicalSubs = state.SubtitleStream != null && !state.SubtitleStream.IsTextSubtitleStream && state.SubtitleDeliveryMethod == SubtitleDeliveryMethod.Encode;
 
+            // If double rate deinterlacing is enabled and the input framerate is 30fps or below, otherwise the output framerate will be too high for many devices
+            var doubleRateDeinterlace = (options.DeinterlaceDoubleRate && (videoStream?.RealFrameRate ?? 60) <= 30);
```
Comment (cvium) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:2092]:
```suggestion
            var doubleRateDeinterlace = options.DeinterlaceDoubleRate && (videoStream?.RealFrameRate ?? 60) <= 30;
```

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 2]
```diff
@@ -2134,35 +2137,36 @@ private string GetFixedSizeScalingFilter(Video3DFormat? threedFormat, int reques
...
                 }
             }
 
             // Add software deinterlace filter before scaling filter
-            if (state.DeInterlace("h264", true)
+            if ((state.DeInterlace("h264", true)
                 || state.DeInterlace("avc", true)
                 || state.DeInterlace("h265", true)
                 || state.DeInterlace("hevc", true))
+                && (!isVaapiH264Encoder && !isQsvH264Encoder && !isNvdecH264Decoder))
```
Comment (cvium) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:2153]: Not sure which is better. Maybe @Bond-009 has an opinion.
```suggestion
                && !isVaapiH264Encoder
                && !isQsvH264Encoder
                && !isNvdecH264Decoder)
```

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs (hunk 1 above)
Comment (orryverducci) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: Good catch, these were accidentally left in after moving it out of an if block. I've corrected it. Thanks!

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs (hunk 2 above)
Comment (orryverducci) [reply MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs]: I've made this change.

File: MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs [hunk 3]
```diff
@@ -2134,35 +2137,38 @@ private string GetFixedSizeScalingFilter(Video3DFormat? threedFormat, int reques
...
+                            doubleRateDeinterlace ? "field" : "frame"));
                 }
             }
 
             // Add software deinterlace filter before scaling filter
-            if (state.DeInterlace("h264", true)
+            if ((state.DeInterlace("h264", true)
                 || state.DeInterlace("avc", true)
                 || state.DeInterlace("h265", true)
                 || state.DeInterlace("hevc", true))
```
Comment (cvium) [inline MediaBrowser.Controller/MediaEncoding/EncodingHelper.cs:2152]: Missed this the first time
```suggestion
                 || state.DeInterlace("avc", true)
                 || state.DeInterlace("h265", true)
                 || state.DeInterlace("hevc", true))
```

**Review by cvium — APPROVED** [review]

**Pull request merged**